doc/design/Double_Buffering.md

```diff
@@ -0,0 +1,89 @@
+Another feasible way is to adopt [staging area](https://en.wikipedia.org/wiki/Staging_(data)). Staging area is an intermediate storage area used for data processing during the [extract, transform and load (ETL)](https://en.wikipedia.org/wiki/Extract,_transform,_load) process. The data staging area sits between the data source(s) and the data target(s), which are often data warehouses, data marts, or other data repositories.   
+The staging area has fixed-size buffer and two methods, `put` and `get`. `put` and `get` access buffer in a thread safe way. The operation of `put` includes copying a data from CPU to GPU and putting data into buffer, and `get` removes a data from buffer.   
+When the program runs, the staging area should be warmed up. First, the `put` is called to put some data in the staging area in advance. After that, the `get` is called every time when the model is calculated, and a data is extracted from the staging area. At the same time, `put` is called, and a data is placed in the staging area. Because the data transfer is used asynchronous, so the `put` method will back immediately.
+For single GPU, the program only has one staging area, this is very simple. For multiple GPU, the program have multiple staging area, and the number of staging area equals to the number of GPU. If model training uses data parallelism, assuming that there is a n fast GPU, `put` divides the data evenly into n portions and copys to the corresponding staging area.
+
+The following C++ programs shows the structure of buffer:
+```
+using MetaType = LoDTensor;
+using BufferElement = std::vector<MetaType>;
+
+class Buffer {
+ private:
+  std::size_t capacity_;
+  std::size_t bytes_limit_;
+  std::size_t current_bytes_;
+  std::mutex mu_;
+  std::condition_variable empty_cond_var_;
+  std::condition_variable full_cond_var_;
+  std::map<paddle::Place ,std::deque<BufferList>> buf_;
+
+ public:
+  void Put(BufferElement* tuple) {...}
+  void Get(BufferElement* tuple) {...}
+  size_t Size() {...}
+  void Clear() {...}
+};
+
+```
+
+**Take the [recognize_digits_mlp](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/tests/book/test_recognize_digits_mlp.py) as an example to show the model definition after using the staging area mechanism.**
+
+```
+image = fluid.layers.data(name='x', shape=[784], dtype='float32')
+label = fluid.layers.data(name='y', shape=[1], dtype='int64')
+
+stage = fluid.staging(input=[image, label])
+stage_program = fluid.default_main_program().clone()
+image, label = stage_out = fluid.unstaging(stage)
+
+y_predict = fluid.layers.fc(input=image, size=1, act=None)
+cost = fluid.layers.square_error_cost(input=y_predict, label=label)
+avg_cost = fluid.layers.mean(x=cost)
+...
+place = fluid.CUDAPlace(0)
+exe = fluid.Executor(place)
+feeder = fluid.DataFeeder(feed_list=[image, label], place=place)
+exe.run(fluid.default_startup_program())
+
+buffer_size = 2
+for pass_id in range(5):
+    for i in range(buffer_size):
```

Comment:
From this API, we can not see that `stage_program` will run in a separated thread.

Reply:
CUDA can use stream mechanism to overlap kernel execution with data transfers. 
https://devblogs.nvidia.com/parallelforall/how-overlap-data-transfers-cuda-cc/

The description here maybe not very clear. I will refine this description.